### PR TITLE
Add types

### PR DESCRIPTION
**Pull request by vanshg**

### CODE REVIEW

File: smile_id_core/__init__.py [hunk 1]
```diff
@@ -2,7 +2,7 @@
 try:
     import importlib.metadata as importlib_metadata
 except ModuleNotFoundError:
-    import importlib_metadata
+    import importlib_metadata  # type: ignore
```
Comment (mldangelo) [inline smile_id_core/__init__.py]: potentially relevant: https://github.com/python-poetry/poetry/blob/c4b2253793cd6b41a99e25e479e40b776cca0a0e/src/poetry/utils/_compat.py#L8-L14

File: smile_id_core/WebApi.py [hunk 2]
```diff
@@ -20,28 +21,34 @@
 
 
 class WebApi:
-    def __init__(self, partner_id, call_back_url, api_key, sid_server):
+    def __init__(
+        self,
+        partner_id: str,
+        call_back_url: str,
+        api_key: str,
+        sid_server: Union[str, int],
```
Comment (mldangelo) [inline smile_id_core/WebApi.py:29]: Consider using a typevar here 
```python
T = TypeVar("T", int, str)
```

File: smile_id_core/IdApi.py [hunk 3]
```diff
@@ -14,23 +16,23 @@
...
 
 class IdApi:
-    def __init__(self, partner_id: str, api_key: str, sid_server: str or int):
+    def __init__(self, partner_id: str, api_key: str, sid_server: str | int):
         if not partner_id or not api_key:
             raise ValueError("partner_id or api_key cannot be null or empty")
         self.partner_id = partner_id
         self.api_key = api_key
         if sid_server in [0, 1, "0", "1"]:
             self.url = sid_server_map[int(sid_server)]
```
Comment (mldangelo) [inline smile_id_core/IdApi.py:25]: consider typing these too. https://mypy.readthedocs.io/en/stable/class_basics.html is a useful guide.

File: smile_id_core/IdApi.py [hunk 4]
```diff
@@ -59,7 +61,9 @@ def submit_job(
             )
         return response
 
-    def __configure_json(self, partner_params, id_params, sec_key):
+    def __configure_json(
+        self, partner_params: dict, id_params: dict, sec_key: dict
```
Comment (mldangelo) [inline smile_id_core/IdApi.py]: We should talk about whether or not it's worthwhile to use things like typeddict for things like partner_params (along with https://pypi.org/project/typing-extensions/ for backports).

File: smile_id_core/__init__.py (hunk 1 above)
Comment (vanshg) [reply smile_id_core/__init__.py]: Ah, indeed. Looks like mypy can understand the sys.version_info check (https://github.com/python/mypy/issues/1393#issuecomment-560165876)

File: smile_id_core/IdApi.py (hunk 3 above)
Comment (vanshg) [reply smile_id_core/IdApi.py:25]: Mypy determines the type via type inference here (i.e. hence the need to cast `sid_server` to a `str` below)

File: smile_id_core/WebApi.py (hunk 2 above)
Comment (vanshg) [reply smile_id_core/WebApi.py:29]: Open to discussion on this -- TypeVar seems like it is meant to be representative of Generics wherein you need to bind multiple parameters to the same type or need to express bounds. Given that we are not doing anything more complex besides initialization of `url` it doesn't seem like the right use case

File: smile_id_core/IdApi.py (hunk 4 above)
Comment (vanshg) [reply smile_id_core/IdApi.py]: It makes sense, possibly in another PR? I think it first makes sense to get us to a state where the overarching types are defined (i.e. dict) and then define the properties explicitly after that

**Review by mldangelo — APPROVED** [review]
Thank you, please update the changelog before you merge this (adding a note to the unreleased section is fine).

**Pull request merged**